File: 495.cpp
#include <bits/stdc++.h>

using namespace std;

typedef  long long int ll;



void multiply(int F[2][2], int M[2][2]);

void power(int F[2][2], int n);

/* function that returns nth Fibonacci number */
int fib(int n)
{
  int F[2][2] = {{1,1},{1,0}};
  if (n == 0)
    return 0;
  power(F, n-1);
  return F[0][0];
}

/* Optimized version of power() in method 4 */
void power(int F[2][2], int n)
{
  if( n == 0 || n == 1)
      return;
  int M[2][2] = {{1,1},{1,0}};

  power(F, n/2);
  multiply(F, F);

  if (n%2 != 0)
     multiply(F, M);
}

void multiply(int F[2][2], int M[2][2])
{
  int x =  F[0][0]*M[0][0] + F[0][1]*M[1][0];
  int y =  F[0][0]*M[0][1] + F[0][1]*M[1][1];
  int z =  F[1][0]*M[0][0] + F[1][1]*M[1][0];
  int w =  F[1][0]*M[0][1] + F[1][1]*M[1][1];

  F[0][0] = x;
  F[0][1] = y;
  F[1][0] = z;
  F[1][1] = w;
}


ofstream myfile;

int main ()
{


myfile.open ("write.txt");

int n , m , a ,b ,t ;

string str;

;



while( scanf("%d",&t) == 1){


    int fi = fib(t);

   cout << "The Fibonacci number for " << t <<" is " << fi << endl;

   myfile << "The Fibonacci number for " << t <<" is " << fi << endl;


}



myfile.close();



return 0;
}

File: 495.java
import java.math.BigInteger;
import java.util.Scanner;

class Main {


 
/* function that returns nth Fibonacci number */
	static BigInteger fib(BigInteger n)
	{
	  BigInteger[][] F = {{ BigInteger.valueOf(1) , BigInteger.valueOf(1) },{ BigInteger.valueOf(1) , BigInteger.valueOf(0) }};
	  if (n.equals( BigInteger.valueOf(0) ) )
	    return BigInteger.valueOf(0);
	  power1(F, n.subtract( BigInteger.valueOf(1) )) ;
	  return F[0][0];
	}
 
/* Optimized version of power() in method 4 */
	static void power1(BigInteger[][] F, BigInteger n)
	{
	  if( n.equals( BigInteger.valueOf(0) )  || n.equals( BigInteger.valueOf(1) ) )
	      return;
	  BigInteger[][] M = {{ BigInteger.valueOf(1) , BigInteger.valueOf(1) },{ BigInteger.valueOf(1) , BigInteger.valueOf(0) }};
	 
	  power1(F, n.divide( BigInteger.valueOf(2) ) );
	  multiply(F, F);
	 
	  if ( !(n.remainder( BigInteger.valueOf(2) )).equals(  BigInteger.valueOf( 0 ) ) )
	     multiply(F, M);
	}
	 
	static void multiply(BigInteger[][] F, BigInteger[][] M)
	{
	  BigInteger x =   ( F[0][0].multiply(M[0][0]) ).add( F[0][1].multiply(M[1][0]) );
	  BigInteger y =  ( F[0][0].multiply(M[0][1]) ).add( F[0][1].multiply(M[1][1]) );  
	  BigInteger z =  ( F[1][0].multiply(M[0][0]) ).add( F[1][1].multiply(M[1][0]) ); 
	  BigInteger w =   ( F[1][0].multiply(M[0][1]) ).add( F[1][1].multiply(M[1][1]) ); 
	 
	  F[0][0] = x;
	  F[0][1] = y;
	  F[1][0] = z;
	  F[1][1] = w;
	}

	public static void main(String[] args) {

		// BigInteger[] f = new BigInteger[5001];
		// f[0] = BigInteger.ZERO;
		// f[1] = BigInteger.ONE;
		// for (int i = 0; i <= 4998; i++)
		// 	f[i + 2] = f[i + 1].add(f[i]);

		Scanner scan = new Scanner(System.in);
		while (scan.hasNextInt()) {
			int n = scan.nextInt();
			System.out.println("The Fibonacci number for " + n + " is " + fib( BigInteger.valueOf( n ) ) );
		}
	}

}

